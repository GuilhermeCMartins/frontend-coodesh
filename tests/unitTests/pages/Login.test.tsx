import React, { ReactElement } from "react";
import {
  render,
  screen,
  fireEvent,
  waitFor,
  RenderResult,
} from "@testing-library/react";
import Login from "@/app/page";
import "@testing-library/jest-dom/extend-expect";
import { useRouter } from "next/navigation";
import { ToastContainer } from "react-toastify";
import { act } from "react-dom/test-utils";

jest.mock("next/navigation", () => ({
  useRouter: jest.fn(),
}));

const renderWithToastify = (component: ReactElement): RenderResult => {
  return render(
    <div>
      <ToastContainer />
      {component}
    </div>
  );
};

const pushMock = jest.fn();

jest.mock("../../../src/app/hooks/useAutentication", () => ({
  __esModule: true,
  default: jest.fn(() => ({
    login: jest.fn().mockImplementation((name, password) => {
      if (name === "usuario_incorreto") {
        throw Error;
      }
      return Promise.resolve({
        success: true,
      });
    }),
  })),
}));

describe("Login page", () => {
  beforeEach(() => {
    jest.clearAllMocks();
    jest.useFakeTimers();

    const useRouter = jest.spyOn(require("next/navigation"), "useRouter");

    useRouter.mockImplementation(() => ({
      query: {},
      push: pushMock,
    }));
  });

  it("matches snapshot", () => {
    const { asFragment } = render(<Login />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render the login form", () => {
    render(<Login />);
    const nameInput = screen.getByLabelText("Nome");
    const passwordInput = screen.getByLabelText("Senha");
    const loginButton = screen.getByText("Entrar");

    expect(nameInput).toBeInTheDocument();
    expect(passwordInput).toBeInTheDocument();
    expect(loginButton).toBeInTheDocument();
  });

  it("should login successfully", async () => {
    renderWithToastify(<Login />);
    const nameInput = screen.getByLabelText("Nome");
    const passwordInput = screen.getByLabelText("Senha");
    const loginButton = screen.getByText("Entrar");

    act(() => {
      fireEvent.change(nameInput, { target: { value: "usuario123" } });
      fireEvent.change(passwordInput, { target: { value: "senha123" } });
    });

    act(() => {
      fireEvent.click(loginButton);
    });

    await waitFor(() =>
      expect(screen.getByText("Login bem sucedido!")).toBeInTheDocument()
    );
  });

  it("should fail to login with incorrect credentials", async () => {
    renderWithToastify(<Login />);
    const nameInput = screen.getByLabelText("Nome");
    const passwordInput = screen.getByLabelText("Senha");
    const loginButton = screen.getByText("Entrar");

    act(() => {
      fireEvent.change(nameInput, { target: { value: "usuario_incorreto" } });
      fireEvent.change(passwordInput, { target: { value: "senha_incorreta" } });
    });

    act(() => {
      fireEvent.click(loginButton);
    });

    await waitFor(() =>
      expect(
        screen.getByText("Usu√°rio ou senha incorretos.")
      ).toBeInTheDocument()
    );
  });
});
