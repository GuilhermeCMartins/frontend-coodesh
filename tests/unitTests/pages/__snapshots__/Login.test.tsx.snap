// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login page matches snapshot 1`] = `
<DocumentFragment>
  <section
    class="c-csXaew"
  >
    <div
      class="MuiContainer-root MuiContainer-maxWidthSm css-cuefkz-MuiContainer-root"
      style="display: flex; flex-direction: column; justify-content: center; align-items: center; box-shadow: 0 2px 4px rgba(0, 0, 0, 0.1); padding: 2rem; border-radius: 8px; background-color: white; min-height: 40vh; border-top: 4px solid #2196f3;"
    >
      <h6
        class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-18k87ye-MuiTypography-root"
        style="margin-bottom: 1rem; text-transform: uppercase; font-weight: 700;"
      >
        Login
      </h6>
      <div
        class="MuiContainer-root MuiContainer-maxWidthXs css-nzinn5-MuiContainer-root"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          style="margin-bottom: 1rem;"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for=":r0:"
            id=":r0:-label"
          >
            Nome
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r0:"
              placeholder="Digite seu nome"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  Nome
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          style="margin-bottom: 1rem;"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for=":r1:"
            id=":r1:-label"
          >
            Senha
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r1:"
              placeholder="Digite sua senha"
              type="password"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  Senha
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth css-1fu7jd5-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Entrar
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        style="margin-top: 1rem;"
      >
        <span
          class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
        >
          Não tem uma conta? 
        </span>
        <a
          href="/register"
          style="text-decoration: none; color: gray;"
        >
          Registre-se
        </a>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
